### PR TITLE
Make Object a copy-on-write datastructure

### PR DESCRIPTION
**Pull request by mystenmark**

### CODE REVIEW

File: crates/sui-types/src/object.rs [hunk 1]
```diff
@@ -1194,6 +1239,32 @@ impl Display for PastObjectRead {
...
+            0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 40, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 123, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+        ]
+    );
+
+    let objref = format!("{:?}", o.compute_object_reference());
+    assert_eq!(objref, "(0x0000000000000000000000000000000000000000000000000000000000000000, SequenceNumber(1), o#59tZq65HVqZjUyNtD7BCGLTD87N5cpayYwEFrtwR4aMz)");
```
Comment (mystenmark) [inline crates/sui-types/src/object.rs:1265]: I verified that this test passes in main

File: crates/sui-types/src/object.rs [hunk 2]
```diff
@@ -615,7 +616,8 @@ impl Display for Owner {
 }
 
 #[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
-pub struct Object {
+#[serde(rename = "Object")]
```
Comment (lxfind) [inline crates/sui-types/src/object.rs:619]: What does this rename do?

File: crates/sui-types/src/object.rs [hunk 3]
```diff
@@ -726,6 +726,45 @@ impl Object {
...
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for Object {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        Arc::make_mut(&mut self.0)
+    }
+}
```
Comment (lxfind) [inline crates/sui-types/src/object.rs:760]: When do we need this?

**Review by lxfind — COMMENTED** [review]
Have you tried to run the single-node benchmark?

**Review by lxfind — APPROVED** [review]
Overall LGTM

File: crates/sui-types/src/object.rs (hunk 2 above)
Comment (mystenmark) [reply crates/sui-types/src/object.rs:619]: Keeps object digests stable - the digest implementation includes the name of the struct as determined by serde.

File: crates/sui-types/src/object.rs (hunk 3 above)
Comment (mystenmark) [reply crates/sui-types/src/object.rs:760]: Any time we assign through a `&mut Object` reference. It is used in temporary store when we write objects and update the latest transaction.

Comment (lxfind) [reply crates/sui-types/src/object.rs:760]: I wonder if it's better to have an explicit swap API instead of &mut ref

**Pull request merged**